File: lib/languages/typescript/function-definitions.ts
// Named function with type annotations
function add(a: number, b: number): number {
  return a + b // Explicit return
}

// Arrow function
const double = (x: number): number => x * 2

// Function with type checking
function process(input: Result<number, Error>): number {
  if (input.ok) {
    return input.value * 2
  } else {
    return 0
  }
}

// Method chaining (similar to pipes)
function transform(numbers: number[]): number[] {
  return numbers
    .map((x) => x * 2)
    .filter((x) => x > 10)
    .sort((a, b) => a - b)
}

// Type for the example
type Result<T, E> = { ok: true; value: T } | { ok: false; error: E }

File: lib/languages/elixir/function-definitions.ex
# Named function with multiple clauses
def add(a, b) do
  a + b  # Implicit return
end

# One-line function syntax
def double(x), do: x * 2

# Function with pattern matching clauses
def process({:ok, value}), do: value * 2
def process({:error, _}), do: 0

# Private function (module-private)
defp helper(x), do: x + 1

# Using the pipe operator
def transform(numbers) do
  numbers
  |> Enum.map(fn x -> x * 2 end)
  |> Enum.filter(fn x -> x > 10 end)
  |> Enum.sort()
end

# Anonymous functions
multiply = fn a, b -> a * b end
result = multiply.(3, 4)  # 12

# Shorthand anonymous functions (capture syntax)
double = &(&1 * 2)
map_fn = &Enum.map/2  # Reference to Enum.map with arity 2

